clients/application-amqp/src/main/java/org/eclipse/hono/application/client/amqp/ProtonBasedApplicationClient.java

```diff
@@ -247,55 +265,69 @@ public final void disconnect(final Handler<AsyncResult<Void>> completionHandler)
                             ProtonHelper.accepted(delivery, true);
                         }
                     } catch (final Throwable t) {
-                        handleException(t, delivery);
+                        handleMessageHandlerError(t, delivery);
                     }
-                },
-                false,
-                s -> Optional.ofNullable(closeHandler).ifPresent(h -> h.handle(null)))
-            .map(recv -> new MessageConsumer() {
-                @Override
-                public Future<Void> close() {
-                    return recv.close();
-                }
-            });
+                }, 
+                closeHandler);
     }
 
     private Future<MessageConsumer> createAsyncConsumer(
             final String sourceAddress,
             final Function<DownstreamMessage<AmqpMessageContext>, Future<Void>> messageHandler,
             final Handler<Throwable> closeHandler) {
 
-        return GenericReceiverLink.create(
-                connection,
+        return createConsumer(
                 sourceAddress,
                 (delivery, message) -> {
                     try {
                         final var msg = ProtonBasedDownstreamMessage.from(message, delivery);
                         messageHandler.apply(msg)
-                            .onSuccess(ok -> {
-                                if (!delivery.isSettled()) {
-                                    LOG.debug("client provided message handler did not settle message, auto-accepting ...");
-                                    ProtonHelper.accepted(delivery, true);
-                                }
-                            })
-                            .onFailure(t -> {
-                                handleException(t, delivery);
-                            });
+                                .onSuccess(ok -> {
+                                    if (!delivery.isSettled()) {
+                                        LOG.debug("client provided message handler did not settle message, auto-accepting ...");
+                                        ProtonHelper.accepted(delivery, true);
+                                    }
+                                })
+                                .onFailure(t -> handleMessageHandlerError(t, delivery));
                     } catch (final Throwable t) {
-                        handleException(t, delivery);
+                        handleMessageHandlerError(t, delivery);
                     }
                 },
-                false,
-                s -> Optional.ofNullable(closeHandler).ifPresent(h -> h.handle(null)))
-            .map(recv -> new MessageConsumer() {
-                @Override
-                public Future<Void> close() {
-                    return recv.close();
-                }
-            });
+                closeHandler);
+    }
+
+    private Future<MessageConsumer> createConsumer(
```

Comment:
can we move this to right after the other private *createConsumer* method?

Reply:
done